### ISSUE TITLE
Custom errors: please add `From<Err<T>> -> Err<U> when U: From<T>`

### ISSUE THREAD

**Issue by chifflier** [Geal/nom#1010]
Nom 5 makes using custom errors better, thanks for that!

Howewer, If you have crate A using `ErrorA` (with trait `ParseError`) and crate B using functions from A, it is fails:

```323 |     parse_x509_der(i)
    |     ^^^^^^^^^^^^^^^^^ expected enum `error::X509Error`, found enum `der_parser::error::BerError`
    |
    = note: expected type `std::result::Result<(&'a [u8], x509::X509Certificate<'a>), nom::Err<error::X509Error>>`
               found type `std::result::Result<(&[u8], x509::X509Certificate<'_>), nom::Err<der_parser::error::BerError>>`

```

And this, even though `X509Error` implements `From<BerError>`.

I think nom could provide a generic solution, adding a generic implementation, depending on the trait implementing `From`. Something lilke:
```rust
impl<T> From<Err<T> for Err<U> when U: From<T> {
// ...
from(t: Err<T>) ->  Err<U> { U::from(t) }
}
```

(not sure of the exact syntax, but you got the idea)
Unfortunately, I can't do that in my crate (rustc complains you can't implement traits on foreign types, so this has to be added in nom).

This would allow nice chaining of custom error types.

**Comment by chifflier** [comment]
Looking further, it seens [nom::Err](https://docs.rs/nom/5.0.0/nom/enum.Err.html#blanket-implementations) implements `Into`.

However, it does not automatically promotes error types, so I had to add a custom combinator, like (simplified version) the following:
```rust
#[macro_export]
macro_rules! upgrade_error (
    ($i:expr, $f:ident) => ({
        match $f($i) {
            Ok(o) => Ok(o),
            Err(::nom::Err::Error(e)) => Err(::nom::Err::Error(e.into())),
            Err(::nom::Err::Failure(e)) => Err(::nom::Err::Failure(e.into())),
            Err(::nom::Err::Incomplete(i)) => Err(::nom::Err::Incomplete(i)),
        }
    });
// ...
);
```

Adding such a combinator to nom would be great!

**Comment by chifflier** [comment]
Last remark: I know it looks similar to `fix_error`, but for some reasonthe above macro works while `fix_error` does not.

**Comment by jonhoo** [comment]
To add to this: this `impl` is pretty much essentially to get ergonomic errors when using a custom nom error type that works with `?`

**Comment by Geal** [comment]
ok so I took a look, and unfortunately, it is not possible to have `impl<T, U: From<T>> From<Err<T> for Err<U>`, because if T and U are the same, it is in conflict with the blanket implementation `impl<T> From<T> for T`.
The alternative would be to automatically convert errors in every combinator, but the type signatures will get complex(and I'm not sure the compiler will easily resolve the types). As an example, here is how I'd rewrite the `pair` combinator, from:

```rust
pub fn pair<I, O1, O2, E: ParseError<I>, F, G>(
  mut first: F, mut second: G,
) -> impl FnMut(I) -> IResult<I, (O1, O2), E> 
where
  F: Parser<I, O1, E>,
  G: Parser<I, O2, E>,
{
  move |input: I| {
    let (input, o1) = first.parse(input)?;
    second.parse(input).map(|(i, o2)| (i, (o1, o2)))
  }
}
```

to:

```rust
pub fn pair<I, O1, O2, E: ParseError<I>+From<E1>+From<E2>, E1: ParseError<I>, E2: ParseError<I>, F, G>(
  mut first: F, mut second: G,
) -> impl FnMut(I) -> IResult<I, (O1, O2), E> 
where
  F: Parser<I, O1, E1>,
  G: Parser<I, O2, E2>,
{
  move |input: I| {
    let (input, o1) = first.parse(input).map_err(Err::convert)?;
    second.parse(input).map(|(i, o2)| (i, (o1, o2))).map_err(Err::convert)
  }
}
```

And it would still not allow the `?` operator :cry:

**Comment by Keruspe** [comment]
What about something like the try! macro we had in the past, but for nom:

```
nom!(parse(input));
```

Which would expand to

```
parse(input).map_err(Err::convert)?;
```

**Comment by Geal** [comment]
yes at this point a simple combinator would be nice. Let's call it `into`, so that it converts the output value and the error if ther's the required `From` implementation? And have the `into()` function combinator, `into!` macro combinator, and `.into()` method on the `Parser` trait. And maybe one day we'll have trait specialization...

**Comment by Geal** [comment]
implemented in 660ce62a7

**Comment by Geal** [comment]
nom 6 is released now :partying_face: 

we'll reopen that issue when(if) specialization lands :)
